fuzz: Make partially_downloaded_block more deterministic

maflcko opened the pull request:
This should make the `partially_downloaded_block` fuzz target even more deterministic.

Follow-up to https://github.com/bitcoin/bitcoin/pull/31841. Tracking issue: https://github.com/bitcoin/bitcoin/issues/29018.

This bundles several changes:

* First, speed up the `deterministic-fuzz-coverage` helper by introducing parallelism.
* Then, a fix to remove spawned test threads or spawn them deterministically. (While testing this, high parallelism and thread contention may be needed)

### Testing

It can be tested via (setting 32 parallel threads):

```
cargo run --manifest-path ./contrib/devtools/deterministic-fuzz-coverage/Cargo.toml -- $PWD/bld-cmake/ $PWD/../b-c-qa-assets/fuzz_corpora/ partially_downloaded_block 32
```

Locally, on a failure, the output would look like:

```diff
 ....
-  150|      0|            m_worker_threads.emplace_back([this, n]() {
-  151|      0|                util::ThreadRename(strprintf("scriptch.%i", n));
+  150|      1|            m_worker_threads.emplace_back([this, n]() {
+  151|      1|                util::ThreadRename(strprintf("scriptch.%i", n));
 ...
```

This excerpt likely indicates that the script threads were started after the fuzz init function returned.

Similarly, for the scheduler thread, it would look like:

```diff
 ...
   227|      0|        m_node.scheduler = std::make_unique<CScheduler>();
-  228|      1|        m_node.scheduler->m_service_thread = std::thread(util::TraceThread, "scheduler", [&] { m_node.scheduler->serviceQueue(); });
+  228|      0|        m_node.scheduler->m_service_thread = std::thread(util::TraceThread, "scheduler", [&] { m_node.scheduler->serviceQueue(); });
   229|      0|        m_node.validation_signals =
 ...
```

Commit:
contrib: Add optional parallelism to deterministic-fuzz-coverage

Co-Authored-By: Hodlinator <172445034+hodlinator@users.noreply.github.com>

Commit:
contrib: Use -Xdemangler=llvm-cxxfilt in deterministic-*-coverage

This makes the result more readable.

Commit:
test: Avoid script check worker threads while fuzzing

Threads may execute their function any time after they are spawned, so
coverage could be non-deterministic.

Fix this,

* for the script check worker threads by disabling them while fuzzing.
* for the scheduler thread by waiting for it to fully start and run the
  service queue.

Prabhat1308 reviewed (commented):
tACK [`fa1e299`](https://github.com/bitcoin/bitcoin/pull/32158/commits/fa1e2995d9996641e79f92549e99a6b37e36d140)

Tested 10 runs with each 32 and 128 parallel threads on MacOS.
Steps followed 
```
cmake -B build -DCMAKE_C_COMPILER="$(brew --prefix llvm)/bin/clang" \ 
   -DCMAKE_CXX_COMPILER="$(brew --prefix llvm)/bin/clang++" \
   -DCMAKE_C_FLAGS="-fprofile-instr-generate -fcoverage-mapping" \
   -DCMAKE_CXX_FLAGS="-fprofile-instr-generate -fcoverage-mapping" \
   -DBUILD_FOR_FUZZING=ON
   
   cmake --build build -j$(sysctl -n hw.ncpu) 
   
   cargo run --manifest-path ./contrib/devtools/deterministic-fuzz-coverage/Cargo.toml -- $PWD/build/ $PWD/qa-assets/fuzz_corpora/ partially_downloaded_block 32/128
```

hodlinator commented on contrib/devtools/deterministic-fuzz-coverage/src/main.rs:
```diff
@@ -187,20 +201,52 @@ The coverage was not deterministic between runs.
-    }
+        Ok(())
+    };
+    thread::scope(|s| -> AppResult {
+        let mut handles = VecDeque::with_capacity(par);
+        let mut res = Ok(());
+        let set_err = |res: &mut _, handle: ScopedJoinHandle<_>| {
+            let r = match handle.join() {
+                Err(_e) => Err("a scoped thread panicked".to_string()),
+                Ok(r) => r,
+            };
+            if r.is_err() {
+                *res = r;
+            }
+        };
+        for (i, entry) in entries.iter().enumerate() {
```
Found a way to reduce this code by 14 lines. What do you think about f8a0a32280d7636135f6821401d7f3b18d10476b?

hodlinator replied:
> I don't think your suggestion works.

I've verified it works by executing something that fails determinism:
```
cargo run --manifest-path ./contrib/devtools/deterministic-fuzz-coverage/Cargo.toml -- $PWD/build_fuzz ../qa-assets/fuzz_corpora/ connman 16
```

> it may return early and leave some threads dangling.

https://doc.rust-lang.org/stable/std/thread/struct.ScopedJoinHandle.html states:

"An owned permission to join on a scoped thread (block on its termination)."

> However, the main thread may now sometimes return an Err, or panic, when a child panics.

The change doesn't modify how panics are handled, it still has `Err("a scoped thread panicked".to_string())`.

> Also, if in the future all errors want to be collected and returned, one would have to revert your change as well?

I don't think the diff would be significantly larger when changing the current version of this PR.

hodlinator replied:
Thanks for the correction.

I believe this would work, even though rustfmt makes the line count less of a win: 67e6d9ece87606297865761f0fb222db9604c180

Commit:
contrib: Only print fuzz output on failure

This makes it humanly possible to track progress as only "[N/M]"-lines are printed as long as we succeed.

Also, use char (a, b) to indicate run_id instead of u8 (0, 1).
Also, use emojis to indicate final success or error.

Co-Authored-By: Hodlinator <172445034+hodlinator@users.noreply.github.com>

hodlinator commented on contrib/devtools/deterministic-fuzz-coverage/src/main.rs:
```diff
@@ -110,8 +120,10 @@ fn deterministic_coverage(
+        let cov_txt_path = build_dir.join(format!("fuzz_det_cov.show.t{thread_id}.r{run_id}.txt"));
+        let profraw_file = build_dir.join(format!("fuzz_det_cov.t{thread_id}.r{run_id}.profraw"));
+        let profdata_file = build_dir.join(format!("fuzz_det_cov.t{thread_id}.r{run_id}.profdata"));
         if !{
             {
                 let mut cmd = Command::new(fuzz_exe);
```
Appreciate you adding the "[N/M]" output, but it's spammed away by fuzz-output even on success. Suggest capturing stdio/stderr and only output them on failure: ea3e89e48250013ea818abcefd7be8e72d31f23d

hodlinator replied:
> if you want less libfuzzer output, it would be easier to compile without libfuzzer

Given the defaults in contrib/devtools/README.md I still think this would be a good change.

Missed that `output()` still allowed checking the exit code. You are right in that it cleaned things up: a7f4e4b7cd0a04ae7c84ad7afa284006a600e369

hodlinator reviewed (commented):
Concept ACK fa1e2995d9996641e79f92549e99a6b37e36d140

janb84 reviewed (commented):
Concept ACK [fa74244](https://github.com/bitcoin/bitcoin/commit/fa74244e97079658a51a808e1efb286fba0fd541)

-Tested, with 8,32 and 128 ✅ (8 will saturate most of the available cores already)

hodlinator reviewed (commented):
ACK 9999358962ebbff7e2821db7081a8c782b0179c8

### Concept

#### Fixes fuzz/unit test nondeterminism

* Waits service thread/scheduler to start up at beginning of tests.
* Disables worker threads for CCheckQueue/script checking.

#### Contrib utilities

* Changes checking for determinism over a multitude of fuzz inputs from a single-threaded slug into an optionally highly parallel affair.
* Limits output from `fuzz` and adds `[N/M]` progress, so humans can get a feeling for when testing would complete (if successful for all inputs).
* Demangles symbols in coverage output.

### Testing

Tested consistently failing determinism output through:
```
cargo run --manifest-path ./contrib/devtools/deterministic-fuzz-coverage/Cargo.toml -- $PWD/build_fuzz ../qa-assets/fuzz_corpora/ connman 16
```

I'm having issues reproducing diffs containing `m_service_thread|scriptch` without the C++ fix. Managed to produce a `scriptch`-containing diff 1-2 times out of ~40 runs. **Any further tips on which qa-assets commit to use, or any other conditions that might help are welcome.** I've seen it succeed multiple times with parallelism of 400, when it fails, I typically don't see those strings. This is the only thing preventing me from ACKing the last commit.

hodlinator commented on src/test/util/setup_common.cpp:
```diff
@@ -60,6 +60,7 @@
 #include <walletinitinterface.h>
 
```
(Inline comment thread to not spam main)

> What is your clang version and do you mind sharing the cmake configure command and the full diff? I just used the tip of qa-assets.

```
₿ cmake -B build_fuzz
CMake Warning at CMakeLists.txt:216 (message):
  BUILD_FOR_FUZZING=ON will disable all other targets and force
  BUILD_FUZZ_BINARY=ON.



Configuring secp256k1 subtree...
-- Could NOT find Valgrind (missing: Valgrind_INCLUDE_DIR Valgrind_WORKS) 


secp256k1 configure summary
===========================
Build artifacts:
  library type ........................ Static
Optional modules:
  ECDH ................................ OFF
  ECDSA pubkey recovery ............... ON
  extrakeys ........................... ON
  schnorrsig .......................... ON
  musig ............................... OFF
  ElligatorSwift ...................... ON
Parameters:
  ecmult window size .................. 15
  ecmult gen table size ............... 86 KiB
Optional features:
  assembly ............................ x86_64
  external callbacks .................. OFF
Optional binaries:
  benchmark ........................... OFF
  noverify_tests ...................... OFF
  tests ............................... OFF
  exhaustive tests .................... OFF
  ctime_tests ......................... OFF
  examples ............................ OFF

Cross compiling ....................... FALSE
Valgrind .............................. OFF
Preprocessor defined macros ........... ENABLE_MODULE_ELLSWIFT=1 ENABLE_MODULE_SCHNORRSIG=1 ENABLE_MODULE_EXTRAKEYS=1 ENABLE_MODULE_RECOVERY=1 ECMULT_WINDOW_SIZE=15 COMB_BLOCKS=43 COMB_TEETH=6 USE_ASM_X86_64=1
C compiler ............................ Clang 19.1.7, /nix/store/ls0g67nsklb2vn3vc9dnksa1adfgq32a-clang-wrapper-19.1.7/bin/clang
CFLAGS ................................ -ftrivial-auto-var-init=pattern
Compile options ....................... -pedantic -Wall -Wcast-align -Wconditional-uninitialized -Wextra -Wnested-externs -Wno-long-long -Wno-overlength-strings -Wno-unused-function -Wreserved-identifier -Wshadow -Wstrict-prototypes -Wundef
Build type:
 - CMAKE_BUILD_TYPE ................... RelWithDebInfo
 - CFLAGS ............................. -O2 -g 
 - LDFLAGS for executables ............ 
 - LDFLAGS for shared libraries ....... 
SECP256K1_APPEND_CFLAGS ............... -fsanitize=undefined,address,fuzzer
SECP256K1_APPEND_LDFLAGS .............. -fsanitize=undefined,address,fuzzer



Configure summary
=================
Executables:
  bitcoind ............................ OFF
  bitcoin-node (multiprocess) ......... OFF
  bitcoin-qt (GUI) .................... OFF
  bitcoin-gui (GUI, multiprocess) ..... OFF
  bitcoin-cli ......................... OFF
  bitcoin-tx .......................... OFF
  bitcoin-util ........................ OFF
  bitcoin-wallet ...................... OFF
  bitcoin-chainstate (experimental) ... OFF
  libbitcoinkernel (experimental) ..... OFF
Optional features:
  wallet support ...................... ON
   - legacy wallets (Berkeley DB) ..... OFF
  external signer ..................... OFF
  ZeroMQ .............................. OFF
  USDT tracing ........................ OFF
  QR code (GUI) ....................... OFF
  DBus (GUI, Linux only) .............. OFF
Tests:
  test_bitcoin ........................ OFF
  test_bitcoin-qt ..................... OFF
  bench_bitcoin ....................... OFF
  fuzz binary ......................... ON

Cross compiling ....................... FALSE
C++ compiler .......................... Clang 19.1.7, /nix/store/ls0g67nsklb2vn3vc9dnksa1adfgq32a-clang-wrapper-19.1.7/bin/clang++
CMAKE_BUILD_TYPE ...................... RelWithDebInfo
Preprocessor defined macros ........... FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
C++ compiler flags .................... -fprofile-instr-generate -fcoverage-mapping -O2 -g -std=c++20 -fPIC -fdebug-prefix-map=/home/hodlinator/bitcoin/src=. -fmacro-prefix-map=/home/hodlinator/bitcoin/src=. -fsanitize=undefined,address,fuzzer -Wall -Wextra -Wgnu -Wformat -Wformat-security -Wvla -Wshadow-field -Wthread-safety -Wloop-analysis -Wredundant-decls -Wunused-member-function -Wdate-time -Wconditional-uninitialized -Woverloaded-virtual -Wsuggest-override -Wimplicit-fallthrough -Wunreachable-code -Wdocumentation -Wself-assign -Wundef -Wno-unused-parameter -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection
Linker flags .......................... -fprofile-instr-generate -fcoverage-mapping -O2 -g -fsanitize=undefined,address,fuzzer -fstack-protector-all -fcf-protection=full -fstack-clash-protection -Wl,-z,relro -Wl,-z,now -Wl,-z,separate-code -fPIE -pie

NOTE: The summary above may not exactly match the final applied build flags
      if any additional CMAKE_* or environment variables have been modified.
      To see the exact flags applied, build with the --verbose option.

Attempt to harden executables ......... ON
Treat compiler warnings as errors ..... OFF
Use ccache for compiling .............. ON


-- Configuring done (0.4s)
-- Generating done (0.1s)
-- Build files have been written to: /home/hodlinator/bitcoin/build_fuzz
```

hodlinator replied:
Log/diff with above build config: [without_fix_400t.gz](https://github.com/user-attachments/files/19547786/without_fix_400t.gz)

Will try with a fresh build directory. Would make sense that sanitizers change behavior.

hodlinator replied:
> It is not visible in the output, as that is now hidden after your suggestion to hide the input

Maybe one could return the `std::process::Output` objects from `run_single` and pass them into `check_diff` to be printed upon failure?

---

First run with fresh build tree and without the C++ thread fix from this PR failed (no determinism-issue detected).

✨ Second run did produce a diff containing both `scriptch.` and `m_service_thread = ...`. ✨

(Third run again detected no determinism).

✨ Fourth run again produced diff with expected substrings. ✨

hodlinator replied:
Was using `> file 2> file` and the output is a big garbled: [without_fix_400t_libFuzzOutput.gz](https://github.com/user-attachments/files/19552215/without_fix_400t_libFuzzOutput.gz)

Had it output the thread id when printing the fuzz-output (`fuzz success thread <thread-id>`), so one can correlate with `t<thead id>` in the diff filenames.

hodlinator replied:
Result with provided patch:
``../src/test/fuzz/partially_downloaded_block.cpp:49 partially_downloaded_block_fuzz_target: Assertion `g_once' failed.``

So it's within the same process. libFuzzer doesn't take `-runs=1` too seriously. Seems weird that it would repeat when only given 1 input... I guess they don't expect deterministic tests.

<details><summary>Full error</summary>

```
⚠️
fuzz failed!
stdout:

stderr:
INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 921745561
INFO: Loaded 1 modules   (543062 inline 8-bit counters): 543062 [0x563af8979360, 0x563af89fdcb6), 
INFO: Loaded 1 PC tables (543062 PCs): 543062 [0x563af89fdcb8,0x563af9247218), 
/home/hodlinator/bitcoin/build_fuzz/bin/fuzz: Running 1 inputs 1 time(s) each.
Running: ../qa-assets/fuzz_corpora/partially_downloaded_block/4eccfaddc420749f641d81b122425972c4f5108e
../src/test/fuzz/partially_downloaded_block.cpp:49 partially_downloaded_block_fuzz_target: Assertion `g_once' failed.
==91561== ERROR: libFuzzer: deadly signal
    #0 0x563af549e44a in __sanitizer_print_stack_trace (/home/hodlinator/bitcoin/build_fuzz/bin/fuzz+0x192744a)
    #1 0x563af539cb90 in fuzzer::PrintStackTrace() (/home/hodlinator/bitcoin/build_fuzz/bin/fuzz+0x1825b90)
    #2 0x563af5377d16 in fuzzer::Fuzzer::CrashCallback() (.part.0) FuzzerLoop.cpp.o
    #3 0x563af5377dda in fuzzer::Fuzzer::StaticCrashSignalCallback() (/home/hodlinator/bitcoin/build_fuzz/bin/fuzz+0x1800dda)
    #4 0x7f0d6b440f2f  (/nix/store/6q2mknq81cyscjmkv72fpcsvan56qhmg-glibc-2.40-66/lib/libc.so.6+0x40f2f) (BuildId: d9765725d929c713f4481765fa13ed815149985f)
    #5 0x7f0d6b49916b in __pthread_kill_implementation (/nix/store/6q2mknq81cyscjmkv72fpcsvan56qhmg-glibc-2.40-66/lib/libc.so.6+0x9916b) (BuildId: d9765725d929c713f4481765fa13ed815149985f)
    #6 0x7f0d6b440e85 in gsignal (/nix/store/6q2mknq81cyscjmkv72fpcsvan56qhmg-glibc-2.40-66/lib/libc.so.6+0x40e85) (BuildId: d9765725d929c713f4481765fa13ed815149985f)
    #7 0x7f0d6b428939 in abort (/nix/store/6q2mknq81cyscjmkv72fpcsvan56qhmg-glibc-2.40-66/lib/libc.so.6+0x28939) (BuildId: d9765725d929c713f4481765fa13ed815149985f)
    #8 0x563af60b3c74 in assertion_fail(std::basic_string_view<char, std::char_traits<char>>, int, std::basic_string_view<char, std::char_traits<char>>, std::basic_string_view<char, std::char_traits<char>>) /home/hodlinator/bitcoin/build_fuzz/../src/util/check.cpp:34:5
    #9 0x563af5950a77 in bool& inline_assertion_check<true, bool&>(bool&, char const*, int, char const*, char const*) /home/hodlinator/bitcoin/build_fuzz/../src/util/check.h:59:13
    #10 0x563af59db569 in partially_downloaded_block_fuzz_target(std::span<unsigned char const, 18446744073709551615ul>) /home/hodlinator/bitcoin/build_fuzz/../src/test/fuzz/partially_downloaded_block.cpp:49:1
    #11 0x563af5cff4c0 in std::function<void (std::span<unsigned char const, 18446744073709551615ul>)>::operator()(std::span<unsigned char const, 18446744073709551615ul>) const /nix/store/dih8vf5naf93c0wcfxqa9pll3k7iv9bm-gcc-14-20241116/include/c++/14-20241116/bits/std_function.h:591:9
    #12 0x563af5cff4c0 in test_one_input(std::span<unsigned char const, 18446744073709551615ul>) /home/hodlinator/bitcoin/build_fuzz/../src/test/fuzz/fuzz.cpp:86:5
    #13 0x563af5cff4c0 in LLVMFuzzerTestOneInput /home/hodlinator/bitcoin/build_fuzz/../src/test/fuzz/fuzz.cpp:205:5
    #14 0x563af53785d8 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/home/hodlinator/bitcoin/build_fuzz/bin/fuzz+0x18015d8)
    #15 0x563af5378b5e in fuzzer::Fuzzer::TryDetectingAMemoryLeak(unsigned char const*, unsigned long, bool) (/home/hodlinator/bitcoin/build_fuzz/bin/fuzz+0x1801b5e)
    #16 0x563af5353e9f in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) (/home/hodlinator/bitcoin/build_fuzz/bin/fuzz+0x17dce9f)
    #17 0x563af535f904 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/hodlinator/bitcoin/build_fuzz/bin/fuzz+0x17e8904)
    #18 0x563af52a9242 in main (/home/hodlinator/bitcoin/build_fuzz/bin/fuzz+0x1732242)
    #19 0x7f0d6b42a1fd in __libc_start_call_main (/nix/store/6q2mknq81cyscjmkv72fpcsvan56qhmg-glibc-2.40-66/lib/libc.so.6+0x2a1fd) (BuildId: d9765725d929c713f4481765fa13ed815149985f)
    #20 0x7f0d6b42a2b8 in __libc_start_main@GLIBC_2.2.5 (/nix/store/6q2mknq81cyscjmkv72fpcsvan56qhmg-glibc-2.40-66/lib/libc.so.6+0x2a2b8) (BuildId: d9765725d929c713f4481765fa13ed815149985f)
    #21 0x563af5349574 in _start (/home/hodlinator/bitcoin/build_fuzz/bin/fuzz+0x17d2574)

NOTE: libFuzzer has rudimentary signal handlers.
      Combine libFuzzer with AddressSanitizer or similar for better crash reports.
SUMMARY: libFuzzer: deadly signal
```

</details>

Commit:
contrib: Warn about using libFuzzer for coverage check

hodlinator commented on contrib/devtools/deterministic-fuzz-coverage/src/main.rs:
```diff
@@ -94,10 +105,14 @@ fn deterministic_coverage(
     let using_libfuzzer = using_libfuzzer(fuzz_exe)?;
-    let profraw_file = build_dir.join("fuzz_det_cov.profraw");
-    let profdata_file = build_dir.join("fuzz_det_cov.profdata");
+    if using_libfuzzer {
+        println!("Warning: The fuzz executable was compiled with libFuzzer as sanitizer.");
+        println!("This tool may be tripped by libFuzzer misbehavior.");
+        println!("It is recommended to compile without libFuzzer.");
+    }
```
nit: Would be nice to get
```
Check if using libFuzzer ... NO
```
For the negative case. Right now on only gets the `...` and then just the omission of this warning.

hodlinator reviewed (approved):
re-ACK fa513101212327f45965092652f6497aa28362ec

Changes since last review:
* Added warning for libFuzzer.
* Changed run ids from r0/r1 to a/b.

### Tests

Fresh build tree:
```
cmake -B build_fuzz2 -DCMAKE_C_COMPILER='clang' -DCMAKE_CXX_COMPILER='clang++' -DBUILD_FOR_FUZZING=ON -DCMAKE_CXX_FLAGS='-fprofile-instr-generate -fcoverage-mapping'
```
After building, ran:
```
cargo run --manifest-path ./contrib/devtools/deterministic-fuzz-coverage/Cargo.toml -- $PWD/build_fuzz2 ../qa-assets/fuzz_corpora/ partially_downloaded_block 400
```
With qa-assets at tip (3846351...).

✨ 2/4 runs without the C++ fix fail with expected differences in scheduler + script check.
✨ 5/5 runs with the C++ fix run deterministically.

---

For concept thoughts, see [previous review](https://github.com/bitcoin/bitcoin/pull/32158#pullrequestreview-2730365813).

janb84 commented on contrib/devtools/deterministic-fuzz-coverage/src/main.rs:
```diff
@@ -187,20 +205,46 @@ The coverage was not deterministic between runs.
+        for (i, entry) in entries.iter().enumerate() {
+            println!("[{}/{}]", i + 1, entries.len());
+            handles.push_back(s.spawn(move || check_individual(entry, i % par)));
+            while handles.len() >= par || i == (entries.len() - 1) || res.is_err() {
```
Thanks for the detailed response! It answered my questions and gave some new insights !

janb84 reviewed (commented):
Re-ACK [fa51310](https://github.com/bitcoin/bitcoin/pull/32158/commits/fa513101212327f45965092652f6497aa28362ec)

Prabhat1308 commented:
re-ACK [`fa51310`](https://github.com/bitcoin/bitcoin/pull/32158/commits/fa513101212327f45965092652f6497aa28362ec)

Tested with various multiple values of parallel threads. All tests passed.
